File: TPS/TP 1/arquivo.java
import java.io.IOException;
import java.io.RandomAccessFile;
import java.util.Scanner;

public class arquivo {

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        int n = sc.nextInt();
        String nomeArquivo = "dados.bin";

        try (RandomAccessFile arquivo = new RandomAccessFile(nomeArquivo, "rw")) {
            for (int i = 0; i < n; i++) {
                double valor = sc.nextDouble();
                arquivo.writeDouble(valor);
            }
        } catch (IOException e) {
            System.err.println("Erro ao escrever no arquivo: " + e.getMessage());
        }

        try (RandomAccessFile arquivo = new RandomAccessFile(nomeArquivo, "r")) {
            long tamanhoArquivo = arquivo.length();
            for (long posicao = tamanhoArquivo - 8; posicao >= 0; posicao -= 8) {
                arquivo.seek(posicao);
                double valor = arquivo.readDouble();
                System.out.println(valor);
            }
        } catch (IOException e) {
            System.err.println("Erro ao ler o arquivo: " + e.getMessage());
        }

        sc.close();
    }
}

File: TPS/TP 1/arquivo.c
#include <stdio.h>
#include <stdlib.h>

int main() {
    int n;
    double valor;
    FILE *arquivo;

    scanf("%d", &n);

 //escrevendo no arquivo
 arquivo = fopen("dados.bin", "wb");
 if (arquivo == NULL) {
     perror("Erro ao abrir o arquivo para escrita");
     return 1;
    }

    for (int i = 0; i < n; i++) {
     scanf("%lf", &valor);
    fwrite(&valor, sizeof(double), 1, arquivo);
    }

  fclose(arquivo);

    //lendo o arquivo de tras pra frente
    arquivo = fopen("dados.bin", "rb");

    fseek(arquivo, 0, SEEK_END);
    long tamanhoArquivo = ftell(arquivo); 

    for (long posicao = tamanhoArquivo - sizeof(double); posicao >= 0; posicao -= sizeof(double)) {
      fseek(arquivo, posicao, SEEK_SET);
      fread(&valor, sizeof(double), 1, arquivo); 
      printf("%.15g\n", valor); 
    }

    fclose(arquivo);

    return 0;
}